### PR TITLE
Introduce Resoruce.Factory (non-reflective api).

### PR DESCRIPTION
**Pull request by serj-lotutovici**
This also opens the path to #103.

### CODE REVIEW

**Comment by serj-lotutovici** [comment]
ping @dhartwich1991 / @Shyish

File: api-client/src/test/java/com/xing/api/XingApiTest.java [hunk 1]
```diff
@@ -57,6 +58,22 @@ public void getRightResourceImpl() throws Exception {
     }
 
     @Test
+    public void getResourceViaFactory() throws Exception {
+        XingApi api = new XingApi.Builder()
+              .custom()
+              .addResourceFactory(new Factory() {
+                  @Override public Resource create(Class<? extends Resource> cls, XingApi api) {
+                      return cls == FactoryResource.class ? new FactoryResource(api, "test") : null;
```
Comment (aballano) [inline api-client/src/test/java/com/xing/api/XingApiTest.java]: You should probably add a test trying to get the wrong factory and checking that the result is null

**Review by dhartwich1991 — APPROVED** [review]
👍

**Review by aballano — CHANGES_REQUESTED** [review]

File: api-client/src/test/java/com/xing/api/XingApiTest.java (hunk 1 above)
Comment (serj-lotutovici) [reply api-client/src/test/java/com/xing/api/XingApiTest.java]: This is guarded by:

``` java
FactoryResource(XingApi api) {
      super(api);
      throw new AssertionError("Resource should be created via factory");
}
```

File: api-client/src/main/java/com/xing/api/resources/UserProfilesResource.java [hunk 2]
```diff
@@ -34,6 +34,12 @@
  * @author daniel.hartwich
  */
 public class UserProfilesResource extends Resource {
+    public static final Resource.Factory FACTORY = new Resource.Factory() {
+        @Override public Resource create(Class<? extends Resource> cls, XingApi api) {
+            return cls == UserProfilesResource.class ? new UserProfilesResource(api) : null;
```
Comment (aballano) [inline api-client/src/main/java/com/xing/api/resources/UserProfilesResource.java]: This code is repeated many times, wouldn't make sense to let Resource implementations just return a new instance and do the class check in the base Resource class?

File: api-client/src/test/java/com/xing/api/XingApiTest.java (hunk 1 above)
Comment (serj-lotutovici) [reply api-client/src/test/java/com/xing/api/XingApiTest.java]: I can add a generic test to the resource tests, but seams redundant.

File: api-client/src/main/java/com/xing/api/resources/UserProfilesResource.java (hunk 2 above)
Comment (serj-lotutovici) [reply api-client/src/main/java/com/xing/api/resources/UserProfilesResource.java]: hm...

File: api-client/src/test/java/com/xing/api/XingApiTest.java (hunk 1 above)
Comment (aballano) [reply api-client/src/test/java/com/xing/api/XingApiTest.java]: I mean something like this test, but calling `api.resource(SomeOtherResourceNotAdded.class)` and checking that the result is null

Comment (serj-lotutovici) [reply api-client/src/test/java/com/xing/api/XingApiTest.java]: `api.resource()` will never return `null` if it can't create a resource via a factory, it will fallback to the reflection api, if that fails it will throw.

File: api-client/src/main/java/com/xing/api/resources/UserProfilesResource.java (hunk 2 above)
Comment (serj-lotutovici) [reply api-client/src/main/java/com/xing/api/resources/UserProfilesResource.java]: @Shyish take another look

**Review by aballano — APPROVED** [review]
